packages/cdktf-cli/lib/cdktf-stack.ts

```diff
@@ -266,83 +132,201 @@ export class CdktfStack {
-      targetAction: "deploy",
-      targetStack: stackName,
-    });
+  private async handleState(cb: () => Promise<void>) {
+    if (this.stopped) {
+      return;
+    }
+
+    try {
+      this.currentWorkPromise = cb();
+      await this.currentWorkPromise;
+      this.notifyState({ type: "done" });
+    } catch (e) {
+      this.currentWorkPromise = undefined;
+      this.notifyState({
+        type: "errored",
+        stackName: this.stack.name,
+        error: String(e),
+      });
+    }
+    this.currentWorkPromise = undefined;
+  }
+
+  public async diff() {
+    await this.handleState(async () => {
+      this.notifyState({ type: "planning", stackName: this.stack.name });
+      const terraform = await this.initalizeTerraform({ isSpeculative: false });
 
-    return this.waitOnMachineDone();
+      const plan = await terraform.plan(false);
+      this.currentPlan = plan;
+      this.notifyState({ type: "planned", stackName: this.stack.name, plan });
+    });
   }
 
-  public async destroy(stackName?: string) {
-    this.stateMachine.send({
-      type: "START",
-      targetAction: "destroy",
-      targetStack: stackName,
+  public async deploy() {
+    await this.handleState(async () => {
+      this.notifyState({ type: "planning", stackName: this.stack.name });
+      const terraform = await this.initalizeTerraform({ isSpeculative: false });
+
+      const plan = await terraform.plan(false);
+      this.notifyState({ type: "planned", stackName: this.stack.name, plan });
+
+      const approved = this.context.autoApprove
+        ? true
+        : await this.waitForApproval(plan);
+
+      if (!approved) {
+        this.notifyState({ type: "dismissed", stackName: this.stack.name });
+        return;
+      }
+
+      this.notifyState({ type: "deploying", stackName: this.stack.name });
+      if (plan.needsApply) {
+        await terraform.deploy(plan.planFile);
+      }
+
+      const outputs = await terraform.output();
+      const outputsByConstructId = getConstructIdsForOutputs(
+        JSON.parse(this.stack.content),
+        outputs
+      );
+
+      this.notifyState({
+        type: "deployed",
+        stackName: this.stack.name,
+        outputs,
+        outputsByConstructId,
+      });
     });
+  }
+
+  public async destroy() {
+    await this.handleState(async () => {
+      this.notifyState({ type: "planning", stackName: this.stack.name });
+      const terraform = await this.initalizeTerraform({ isSpeculative: false });
+
+      const plan = await terraform.plan(true);
+      this.notifyState({ type: "planned", stackName: this.stack.name, plan });
+
+      const approved = this.context.autoApprove
+        ? true
+        : await this.waitForApproval(plan);
+      if (!approved) {
+        this.notifyState({ type: "dismissed", stackName: this.stack.name });
+        return;
+      }
 
-    return this.waitOnMachineDone();
+      this.notifyState({ type: "destroying", stackName: this.stack.name });
```

Comment:
We could `Omit<StackUpdate, 'stackName'> in `notifyState` to get rid of all the `stackName: this.stack.name` in numerous places and add the `stackName` in `notifyState`. What do you think? Would this make the code simpler in your eyes? Or harder to follow / grasp?

Reply:
I tried, but it does not work with our typescript version. The `type` discriminator can not be used anymore to infer the message structure after passed through Omit. I'd rather duplicate stackName then interfering with TS here